firedrake/slate/slate.py

```diff
@@ -0,0 +1,673 @@
+                        it_type = integral.integral_type()
+                        subdata = integral.subdomain_data()
+
+                        data = subdomain_data[domain].get(it_type)
+                        if data is None:
+                            subdomain_data[domain][it_type] = subdata
+                        elif subdata is not None:
+                            assert data.ufl_id() == subdata.ufl_id(), "Integrals in the tensor must have the same subdomain_data objects."
+                self._subdomain_data = subdomain_data
+
+    def ufl_domains(self):
+        """Returns the integration domains of the integrals associated with
+        the tensor."""
+        return self._integral_domains
+
+    def ufl_domain(self):
+        """This function returns a single domain of integration occuring
+        in the tensor.
+
+        The function will fail if multiple domains are found."""
+        domains = self.ufl_domains()
+        assert all(domain == domains[0] for domain in domains), "All integrals must share the same domain of integration."
+
+        return domains[0]
+
+    def subdomain_data(self):
+        """Returns a mapping on the tensor:
+        `{domain:{integral_type: subdomain_data}}`."""
+        return self._subdomain_data
+
+    def __hash__(self):
+        """Returns a hash code for use in dictionary objects."""
+        if self._hash is None:
+            self._hash = hash((type(self), )
+                              + tuple(hash(a) for a in self.arguments())
+                              + tuple(hash(c) for c in self.coefficients()))
+
+        return self._hash
+
+
+class Scalar(Tensor):
+    """A scalar representation of a 0-form. This class wraps the
+    relevant information provided from a :class:`ufl.Form` object
+    for use in symbolic linear algebra computations.
+
+    Note that this class will complain if you provide an incompatible
+    form object as input. The `Scalar` class expects a form of rank 0.
+
+    :arg form: a :class:`ufl.Form` object representing a 0-form.
+    """
+
+    def __init__(self, form):
+        """Constructor for the Scalar class."""
+        r = len(form.arguments())
+        if r != 0:
+            raise Exception("Cannot create a `slate.Scalar` from a form with rank %d" % r)
+        self.form = form
+        self.check_integrals(form.integrals())
+        super(Scalar, self).__init__()
+        # Generate integral data after tensor has been initialized
+        self.generate_integral_info(form.integrals())
+
+    def arguments(self):
+        """Returns a tuple of arguments associated with the tensor."""
+        return self.form.arguments()
+
+    def coefficients(self):
+        """Returns a tuple of coefficients associated with the tensor."""
+        return self.form.coefficients()
+
+    def __str__(self, prec=None):
+        """String representation of a SLATE Scalar object."""
+        return "S_%d" % self.id
```

Comment:
So it seems that `Scalar`/`Vector`/`Tensor` exhibit the same behaviour except for this letter here (S/V/M). I suggest merging these classes, and instead simply picking this letter based on the rank.

You may keep `Scalar`/`Vector`/`Tensor` as factory functions if you really feel the need for them.